quinn/src/connection.rs

```diff
@@ -122,20 +158,145 @@ impl Connecting {
     /// Will panic if called after `poll` has returned `Ready`.
     pub fn remote_address(&self) -> SocketAddr {
         match self.0.as_ref().unwrap() {
-            ConnectingState::Handshaking(handshaking) => handshaking.remote_address(),
+            ConnectingState::Incoming(ref incoming) => incoming.inner.remote_address(),
+            ConnectingState::Handshaking(ref handshaking) => handshaking.remote_address(),
+            ConnectingState::AcceptError(_) => {
+                panic!("remote_address called after connecting errored")
+            }
+        }
+    }
+
+    fn unwrap_incoming_ref(&self, caller: &str) -> &ConnectingIncoming {
+        match self.0.as_ref().unwrap() {
+            ConnectingState::Incoming(ref incoming) => incoming,
+            _ => panic!("{} called after already started handshake", caller),
+        }
+    }
+
+    fn unwrap_incoming(mut self, caller: &str) -> ConnectingIncoming {
+        match self.0.take().unwrap() {
+            ConnectingState::Incoming(incoming) => incoming,
+            _ => panic!("{} called after already started handshake", caller),
+        }
+    }
+
+    /// Whether the socket address that is initiating this connection has been validated
+    ///
+    /// This means that the sender of the initial packet has proved that they can receive traffic
+    /// sent to `self.remote_address()`.
+    ///
+    /// Will panic if called when the handshake has already begun.
+    pub fn is_validated(&self) -> bool {
+        self.unwrap_incoming_ref("is_validated")
+            .inner
+            .is_validated()
+    }
+
+    /// Whether it is legal to require the client to retry
+    ///
+    /// If `is_validated` is false, `may_retry` is necessarily true.
+    ///
+    /// Will panic if called when the handshake has already begun or if `may_retry` is false.
+    pub fn may_retry(&self) -> bool {
+        self.unwrap_incoming_ref("may_retry").inner.is_validated()
+    }
+
+    /// Reject this incoming connection attempt
+    ///
+    /// Will panic if called when the handshake has already begun.
+    pub fn reject(self) {
```

Comment:
Let's also have a method to `ignore` a connection attempt.

Reply:
Done.